update .dockercfg content to config.json

juanvallejo opened the pull request:
update the data format of .dockercfg to match the new docker config.json
format, which encapsulates all registry auth objects in an overall
"auths" object when an option `--config-format` is specified with the value
`--config-format=config.json`:

```
{
    "auths": {
        "reg.url": {
            "auth": "...=="
        }
    }
}
```

**Release note**:
```release-note
secret data containing Docker registry auth objects is now generated using the config.json format
```

cc @bparees @fabianofranz

juanvallejo committed:
update .dockercfg content to config.json

update the data format of .dockercfg to match the new docker config.json
format, which encapsulates all registry auth objects in an overall
"auths" object:

{
    "auths": {
        "reg.url": {
            "auth": "...=="
        }
    }
}

bparees commented:
there's an option question from me at least as to whether we can switch to use this format as the default (because it's what most registries expect these days and some fail if it's not in this format), or must we continue using the existing format by default for backwards compatibility. (which while more compatible, is more inconvenient for most users who will have to specify a flag to get the right behavior)

juanvallejo commented:
> there's an option question from me at least as to whether we can switch to use this format as the default (because it's what most registries expect these days and some fail if it's not in this format), or must we continue using the existing format by default for backwards compatibility. (which while more compatible, is more inconvenient for most users who will have to specify a flag to get the right behavior)

cc @fabianofranz @smarterclayton

dims commented:
/ok-to-test

juanvallejo commented:
/assign fabianofranz
/assign smarterclayton

smarterclayton commented:
I don't think there is a version of docker in use today that doesn't
support config.json, and old and new Kube clusters support both.

We should switch to generating the new file and instruct users to use an
older client or write their own if they need the old form.  I don't see the
type switch as necessary.


On Oct 16, 2017, at 7:25 PM, Juan Vallejo <notifications@github.com> wrote:

/assign fabianofranz
/assign smarterclayton

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/53916#issuecomment-336959794>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p3lcngzBTHC7DeFgx1lrkOHkBDxpks5ss5GFgaJpZM4P5FZI>
.

juanvallejo commented:
@smarterclayton 
> We should switch to generating the new file and instruct users to use an
older client or write their own if they need the old form.  I don't see the
type switch as necessary.

Sounds good to me, although I think there might be some value for the user to still have the option to generate the old form, at least with a flag. I could switch to have the `config.json` form be the default, and leave the old form as an option, wdyt?

smarterclayton commented:
I don't know why we would need to generate the old format.  It's no longer
the default, and the secret it creates is valid for end users on
Kubernetes.  The old format is not guaranteed to work forever, so users do
need to start migrating, and creation is a reasonable time.

I don't know whether supporting the flag really helps a user.

On Mon, Oct 16, 2017 at 11:06 PM, Juan Vallejo <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton>
>
> We should switch to generating the new file and instruct users to use an
> older client or write their own if they need the old form. I don't see the
> type switch as necessary.
>
> Sounds good to me, although I think there might be some value for the user
> to still have the option to generate the old form, at least with a flag. I
> could switch to have the config.json form be the default, and leave the
> old form as an option, wdyt?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/53916#issuecomment-337043138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwoBh4dYu2Ss7bFfK6x4fLMDIajoks5ss8U7gaJpZM4P5FZI>
> .
>

juanvallejo commented:
@smarterclayton thanks, went ahead and removed the `config-format` flag added in the previous commit and updated `SecretForDockerRegistryGeneratorV1` to serialize registry auth data in the`config.json` format

juanvallejo commented:
/test pull-kubernetes-kubemark-e2e-gce

derekwaynecarr commented:
squash commits, then lgtm

juanvallejo commented:
@derekwaynecarr thanks, done!

derekwaynecarr commented:
/approve no-issue
/lgtm

juanvallejo commented:
ping @fabianofranz

fabianofranz commented:
/approve no-issue

dearjadu commented:
Existing secrets with `.dockercfg` from earlier versions of kubectl (1.8 for sure) is breaking with this change. In fact, this change is backward incompatible and has come in without depreciation. Fortunately, we are not using this on production. Also, note that `kubectl create secret` by default still creates secrets using the older format, unless the user explicitly sets the format in the `create secret` command.

juanvallejo commented:
> Existing secrets with .dockercfg from earlier versions of kubectl (1.8 for sure) is breaking with this change. In fact, this change is backward incompatible and has come in without depreciation. Fortunately, we are not using this on production. Also, note that kubectl create secret by default still creates secrets using the older format, unless the user explicitly sets the format in the create secret command.

cc @smarterclayton for thoughts
I can go ahead and update `kubectl create secret` to use config.json format

smarterclayton commented:
We should update to create those secrets by default in the new format.




On Dec 18, 2017, at 5:38 PM, Juan Vallejo <notifications@github.com> wrote:

Existing secrets with .dockercfg from earlier versions of kubectl (1.8 for
sure) is breaking with this change. In fact, this change is backward
incompatible and has come in without depreciation. Fortunately, we are not
using this on production. Also, note that kubectl create secret by default
still creates secrets using the older format, unless the user explicitly
sets the format in the create secret command.

cc @smarterclayton <https://github.com/smarterclayton> for thoughts
I can go ahead and update kubectl create secret to use config.json format

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/53916#issuecomment-352580673>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p8KZaZjMR9SGW--kV6Ob9EO-s8Utks5tBul-gaJpZM4P5FZI>
.

liggitt commented:
Did this move the key to config.json? I think putting the new format in the old key is incorrect.

redbaron commented:
Tracked down my problems to this PR just to find out that it was solved 2 days ago (which is good), but OMG, why did this PR change format of .dockercfg?? it is a legacy key and should be immutable, if new format is needed then create new `config.json` but please leave all code related to `.dockercfg` alone.

If new 1.9.x kubectl is meant to work with 1.8.x kubelets, then it should continue to create `.dockercfg` key alongside new `config.json` for foreseeable future.
